utopia-remix/app/util/github-branch-contents.server.spec.ts

```diff
@@ -0,0 +1,287 @@
+                },
+              },
+              directory: { type: 'DIRECTORY' },
+              fullPath: '/some/dir',
+              type: 'PROJECT_CONTENT_DIRECTORY',
+            },
+          },
+          directory: { type: 'DIRECTORY' },
+          fullPath: '/some',
+          type: 'PROJECT_CONTENT_DIRECTORY',
+        },
+      })
+    })
+    it('populates the leaf for a file entry', async () => {
+      let root = projectContentDirectory('root')
+      const target = populateDirectories({
+        root_MUTABLE: root,
+        relativeFilePath: 'some/file/here.json',
+      })
+      await populateEntryContents({
+        filePath: 'some/file/here.json',
+        entry: {
+          type: 'File',
+          vars: { uncompressedSize: 123 },
+          buffer: async () => Buffer.from(`"hello there"`),
+        } as UnzipEntry,
+        target_MUTABLE: target,
+        existingAssets_MUTABLE: [],
+        assetsToUpload_MUTABLE: [],
+      })
+      expect(root.children).toEqual({
+        some: {
+          children: {
+            file: {
+              children: {
+                'here.json': {
+                  content: {
+                    fileContents: {
+                      code: '"hello there"',
+                      parsed: { type: 'UNPARSED' },
+                      revisionsState: 'CODE_AHEAD',
+                    },
+                    lastParseSuccess: null,
+                    lastSavedContents: null,
+                    type: 'TEXT_FILE',
+                    versionNumber: 0,
+                  },
+                  fullPath: '/some/file/here.json',
+                  type: 'PROJECT_CONTENT_FILE',
+                },
+              },
+              directory: { type: 'DIRECTORY' },
+              fullPath: '/some/file',
+              type: 'PROJECT_CONTENT_DIRECTORY',
+            },
+          },
+          directory: { type: 'DIRECTORY' },
+          fullPath: '/some',
+          type: 'PROJECT_CONTENT_DIRECTORY',
+        },
+      })
+    })
+    it('adds a new file to upload if there is a need', async () => {
+      let root = projectContentDirectory('root')
+      const target = populateDirectories({
+        root_MUTABLE: root,
+        relativeFilePath: 'some/file/here.png',
+      })
+
+      const toUpload: AssetToUpload[] = []
+
+      await populateEntryContents({
+        filePath: 'some/file/here.png',
+        entry: {
+          type: 'File',
+          vars: { uncompressedSize: 123 },
+          buffer: async () => Buffer.from(`"hello there"`),
+        } as UnzipEntry,
+        target_MUTABLE: target,
+        existingAssets_MUTABLE: [],
+        assetsToUpload_MUTABLE: toUpload,
+      })
+
+      expect(root.children).toEqual({
+        some: {
+          children: {
+            file: {
+              children: {
+                'here.png': {
+                  content: {
+                    gitBlobSha: '6ed5c1e4f0c1dc6bd7f3e6a0a52c20585f5420be', // <- this!
```

Comment:
What about it?

Reply:
it was just a note I left when writing the test to signal the meaningful things to look for in the contents, since the whole json object is noisy. If it's unwanted I can get rid of it.